ipapython/ipaldap.py

```diff
@@ -1749,3 +1752,239 @@ def get_ldap_uri(host='', port=389, cacert=None, ldapi=False, realm=None,
+
+    def remove_cache_entry(self, dn):
+        assert isinstance(dn, DN)
+        self.emit('DROP: %s', dn)
+        if dn in self.cache:
+            del self.cache[dn]
+        else:
+            self.emit('DROP: not in cache %s', dn)
+
+    #  Begin LDAPClient methods
+
+    def add_entry(self, entry):
+        self.emit('add_entry')
+        self.remove_cache_entry(entry.dn)
+        super(LDAPCache, self).add_entry(entry)
+
+    def update_entry(self, entry):
+        self.emit('update_entry')
+        self.remove_cache_entry(entry.dn)
+        super(LDAPCache, self).update_entry(entry)
+
+    def delete_entry(self, entry_or_dn):
+        self.emit('delete_entry')
+        if isinstance(entry_or_dn, DN):
+            dn = entry_or_dn
+        else:
+            dn = entry_or_dn.dn
+
+        self.remove_cache_entry(dn)
+
+        super(LDAPCache, self).delete_entry(dn)
+
+    def move_entry(self, dn, new_dn, del_old=True):
+        self.emit('move_entry')
+        self.remove_cache_entry(dn)
+        self.remove_cache_entry(new_dn)
+
+        super(LDAPCache, self).move_entry(dn, new_dn, del_old)
+
+    def modify_s(self, dn, modlist):
+        self.emit('modify_s')
+        if not isinstance(dn, DN):
+            dn = DN(dn)
+
+        self.emit('modlist %s', modlist)
+        for (_op, attr, mod_dn) in modlist:
+            if attr.lower() in ('member',
+                                'ipaallowedtoperform_write_keys',
+                                'managedby_host'):
+                for d in mod_dn:
+                    if not isinstance(d, (DN, RDN)):
+                        d = DN(d.decode('utf-8'))
+                        self.emit('modify_s %s', d)
+                        self.remove_cache_entry(d)
+
+        self.emit('modify_s %s', dn)
+        self.remove_cache_entry(dn)
+
+        return super(LDAPCache, self).modify_s(dn, modlist)
+
+    def get_entry(self, dn, attrs_list=None, time_limit=None,
+                  size_limit=None, get_effective_rights=False):
+        # pylint: disable=no-member
+        if not self._enable_cache:
+            return super(LDAPCache, self).get_entry(
+                dn, attrs_list, time_limit, size_limit, get_effective_rights
+            )
+        self.emit("Cache lookup: %s", dn)
+
+        entry = self.cache.get(dn)
+        if get_effective_rights and entry:
+            # We don't cache this so do the query but don't drop the
+            # entry.
+            entry = None
+
+        if entry and entry.exception:
+            hits = self._cache_hits + 1  # pylint: disable=no-member
+            object.__setattr__(self, '_cache_hits', hits)
+            self.emit("HIT: Re-raising %s", entry.exception)
+            self.cache_status('HIT')
+            raise entry.exception
+
+        self.emit("Requested attrs_list %s", attrs_list)
+        if entry:
+            self.emit("Cached attrs_list %s", entry.attrs_list)
+
+        if not attrs_list:
```

Comment:
We have a number of places where attrs_list = [''] which should be equivalent to []/['*']. I'm experimenting with this. I think we can save additional searches.

Reply:
So get_dn() calls generally just do that, get the dn of an entry. This is still cacheable but `if entry `returns False so these were being ignored. Explicitly set the attrs_list to ['dn'] and change the method to see if there is a cache entry.

Also add a log parser I wrote to help calculate the cache hits, misses, etc.